README
DigestX::Multi

"Digest::Many" aims to simplify common work-flows that are necessary
where-ever somebody might wish to compute multiple digests for the same
content, specifically, with care not to need to open the same files for
multiple passes.

The need to do so arises in especially paranoid situations where the
potential concern involves the potential of any one digest algorithm to
be attacked and have a collision built for, based on the notion that
finding one collision for a given algorithm is hard enough, but it aught
to be many orders of magnitude more difficult to generate a collision
that simultaneously collides under two or more algorithms.

Thus, increasing the number of digest algorithms used exponentially
hardens your attack surface, slightly more so than increasing the digest
size of a single algorithm would achieve.

However, this makes the data less useful for applications where a digest
is used to hide/validate some data using a shared secret, as re-using
the same shared secret with the same content under multiple digests
slightly increases the total number of bits that may relate to the
secret, and increases the chance that the secret may be obtained by
breaking a favorable algorithm.

Note that of course, you should not really be using naive secret-prefix
or secret-suffix based message digest authentication, but you should
instead use a proper "HMAC" algorithm
<https://rdist.root.org/2009/10/29/stop-using-unsafe-keyed-hashes-use-hm
ac/>, otherwise you're exposing yourself to a length extension attack
<https://www.happybearsoftware.com/you-are-dangerously-bad-at-cryptograp
hy.html>

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for DigestX-Multi

0.001000 2016-06-18T20:44:58Z 3904dd2
 - First version.


t/lib/KENTNL/DigestXTest.pm
use strict;
use warnings;

package KENTNL::DigestXTest;

use DigestX::Multi;
use Test::Builder;

sub test_algo {
  my ( $class, $label, $algo_gen_sub ) = @_;

  my $md = DigestX::Multi->new();
  $md->add_digester( $label . 'digest-1' => $algo_gen_sub->() );
  $md->append_bytes("Hello World");

  my $hash_1 = $md->digests( { format => 'base64' } );
  my $hash_2 = $md->digests( { format => 'base64' } );

  my $builder = Test::Builder->new();

  $builder->note('Testing digests() is non-destructive');

  return
    unless $builder->ok( exists $hash_1->{ $label . 'digest-1' }, $label . ': digest-1 exists in digest-hash 1' )
    and $builder->ok( exists $hash_2->{ $label . 'digest-1' }, $label . ': digest-1 exists in digest-hash 2' );

  return
    unless $builder->ok( defined $hash_1->{ $label . 'digest-1' }, $label . ': digest-1 defined in digest-hash 1' )
    and $builder->ok( defined $hash_2->{ $label . 'digest-1' }, $label . ': digest-1 defined in digest-hash 2' );

  if ( $hash_1->{ $label . 'digest-1' } eq $hash_2->{ $label . 'digest-1' } ) {
    $builder->ok( 1, $label . ': First digest() matches second digest() for digest-1' );
  }
  else {
    $builder->ok( 0, $label . ': First digest() matches second digest() for digest-1' );
    $builder->diag("Digests do not match for $label");
    $builder->diag( " - Expected: " . $hash_1->{ $label . 'digest-1' } );
    $builder->diag( " - Got: " . $hash_2->{ $label . 'digest-1' } );
    return;
  }

  $builder->note('Testing adding new digesters late in cycle');

  $md->add_digester( $label . 'digest-2', $algo_gen_sub->() );
  $md->append_bytes("Hello World");

  my $hash_3 = $md->digests( { format => 'base64' } );

  return
    unless $builder->ok( exists $hash_3->{ $label . 'digest-1' }, $label . ': digest-1 exists in digest-hash 3' )
    and $builder->ok( exists $hash_3->{ $label . 'digest-2' }, $label . ': digest-2 exists in digest-hash 3' );

  return
    unless $builder->ok( defined $hash_3->{ $label . 'digest-1' }, $label . ': digest-2 defined in digest-hash 3' )
    and $builder->ok( defined $hash_3->{ $label . 'digest-2' }, $label . ': digest-2 defined in digest-hash 3' );

  if ( $hash_3->{ $label . 'digest-1' } ne $hash_3->{ $label . 'digest-2' } ) {
    $builder->ok( 1, $label . ': Early digest-1 does not equal later digest-2' );
  }
  else {
    $builder->ok( 0, $label . ': Early digest-1 does not equal later digest-2' );
    $builder->diag("Digests unexpectedly match for $label");
    $builder->diag(" - Expected: Anything else");
    $builder->diag( " - Got: " . $hash_3->{ $label . 'digest-1' } );
    return;
  }

  $builder->ok( $md->is_valid( { format => 'base64', digests => $hash_3 } ), "$label: Should validate against identical hash" );
  $builder->ok( !$md->is_valid( { format => 'base64', digests => $hash_2 } ), "$label: Should not validate against stale hash" );

  my $copy = $md->clone();

  my $ltest = $builder->ok(
    $md->is_valid( { format => 'base64', digests => $hash_3 } ),
    "$label: Original Should validate against identical hash after a clone"
  );
  undef $ltest
    unless $builder->ok(
    $copy->is_valid( { format => 'base64', digests => $hash_3 } ),
    "$label: Clone Should validate against identical hash after a clone"
    );

  if ( not $ltest ) {
    require Test::More;
    $builder->diag(
      Test::More::explain(
        {
          hash_cmp => $hash_3,
          orig     => [ $md->_validity_map( { format => 'base64', digests => $hash_3 } ) ],
          orig_hash => $md->digests( format => 'base64' ),
          copy      => [ $copy->_validity_map( { format => 'base64', digests => $hash_3 } ) ],
          copy_hash => $copy->digests( format  => 'base64' ),
        }
      )
    );
  }

  $copy->reset_digesters();

  $ltest = $builder->ok( $md->is_valid( { format => 'base64', digests => $hash_3 } ),
    "$label: Original Should validate against identical hash after a clone+reset" );
  undef $ltest
    unless $builder->ok(
    !$copy->is_valid( { format => 'base64', digests => $hash_3 } ),
    "$label: Clone Should not validate against identical hash after a clone+reset"
    );

  if ( not $ltest ) {
    require Test::More;
    $builder->diag(
      Test::More::explain(
        {
          hash_cmp => $hash_3,
          orig     => [ $md->_validity_map( { format => 'base64', digests => $hash_3 } ) ],
          orig_hash => $md->digests( format => 'base64' ),
          copy      => [ $copy->_validity_map( { format => 'base64', digests => $hash_3 } ) ],
          copy_hash => $copy->digests( format  => 'base64' ),
        }
      )
    );
  }

}

1;



lib/DigestX/Multi.pm
use 5.006;    # our
use strict;
use warnings;

package DigestX::Multi;

our $VERSION = '0.001000';
use Carp qw( croak carp );

# ABSTRACT: Simplify digesting the same content with multiple digesters.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY













sub new {
  my ( $class, @args, ) = @_;
  my $arg_hash = { ref $args[0] ? %{ $args[0] } : @args };
  my $self = bless $arg_hash, $class;
  return $self;
}
















sub add_digester {
  my ( $self, $name, $digester, ) = @_;
  croak "Can't add digester with name $name, one already exists" if exists $self->{digesters}->{$name};
  croak "Did not pass object for $name, got undef" unless defined $digester;
  $self->{digesters}->{$name} = $digester;
  return $self;
}










sub clone {
  my ( $self, ) = @_;
  my $class = ref $self;
  my $clone = bless {}, $class;
  for ( keys %{ $self->{digesters} || {} } ) {
    my $clone_digester = $self->{digesters}->{$_}->clone;
    croak $self->{digesters}->{$_} . "->clone() returned undef for $_" if not defined $clone_digester;
    $clone->{digesters}->{$_} = $clone_digester;
  }
  return $clone;
}









sub reset_digesters {
  my ( $self, ) = @_;
  $self->{digesters}->{$_}->reset for keys %{ $self->{digesters} || {} };
  return $self;
}











sub append_bytes {
  my ( $self, @bytes ) = @_;
  $self->{digesters}->{$_}->add(@bytes) for keys %{ $self->{digesters} || {} };
  return $self;
}













sub append_filehandle {
  my ( $self, $filehandle ) = @_;
  my $n;
  my $buf = q[];
  while ( $n = read $filehandle, $buf, 4 * 1024 ) {
    $self->append_bytes($buf);
  }
  return $self;
}











sub append_file {
  my ( $self, $path ) = @_;
  open my $fh, '<:raw', $path or croak "Can't open $path, $!";
  $self->append_filehandle($fh);
  close $fh or carp "Error closing $path";
  return $self;
}

sub _digest_binary   { $_[0] }
sub _digest_hex      { unpack 'H*', $_[0] }
sub _digest_uuencode { pack 'u', $_[0] }

sub _digest_base64 {
  require MIME::Base64;
  ## no critic (ProhibitCallsToUnexportedSubs)
  my $encoded = MIME::Base64::encode( $_[0], q[] );
  $encoded =~ s{=+$}{}sx;
  return $encoded;
}























sub digests {
  my ( $self, @args ) = @_;
  my $arg_hash = { ref $args[0] ? %{ $args[0] } : @args };
  $arg_hash->{format} ||= 'binary';
  $arg_hash->{encode} ||= do {
    my $ref = $self->can( '_digest_' . $arg_hash->{format} );
    croak "Unknown format $arg_hash->{format}" if not defined $ref;
    $ref;
  };
  my $result = {};
  for ( keys %{ $self->{digesters} || {} } ) {
    my $clone = $self->{digesters}->{$_}->clone;
    croak "Clone of $self->{digesters}->{$_} for $_ returned undef" if not defined $clone;
    $result->{$_} = $arg_hash->{encode}->( $clone->digest );
  }
  return $result;
}













sub is_valid {
  my ( $self, @args ) = @_;
  for my $result ( $self->_validity_map(@args) ) {
    return q[] if not $result->{pass};
  }
  return 1;
}













sub assert_valid {
  my ( $self, @args ) = @_;
  #<<<
  my (@errs) = map  { $_->{reason} || 'Undefined Reason' }
               grep { !$_->{pass}                        }
               $self->_validity_map(@args);
  #>>>
  return 1 unless @errs;
  croak "Hash digests failed validation against expected:\n\t-" . ( join qq[,\n\t], @errs ) . "\n at ";
}

# This string comparison is intentionally sub-optimal to avoid timing attacks
# based on the fact a mismatching digest will return faster than a matching one.
sub _teq {
  ## no critic (ProhibitAmbiguousNames)
  my ( $left, $right ) = @_;
  return q[] if not defined $left or not defined $right;
  return q[] if ( length $left ) != ( length $right );
  my (@left)  = map { ord } split //sx, $left;
  my (@right) = map { ord } split //sx, $right;
  my $sum;

  # Counts the number of characters that don't match using XOR
  ## no critic (ProhibitBitwiseOperators)
  $sum += $left[$_] ^ $right[$_] for 0 .. $#left;
  return 0 == $sum;
}

sub _validity_map {
  my ( $self, @args ) = @_;
  my $arg_hash = { ref $args[0] ? %{ $args[0] } : @args };

  croak q[No hash 'digests' to validate] unless exists $arg_hash->{'digests'};

  my %want_hash = %{ delete $arg_hash->{'digests'} };
  my %got_hash  = %{ $self->digests($arg_hash) };

  my %all_keys = map { ( $_ => 1 ) } keys %got_hash, keys %want_hash;

  my @results;

  push @results, { pass => 0, reason => 'HASH:INPUT:EMPTY' }    if not keys %got_hash;
  push @results, { pass => 0, reason => 'HASH:EXPECTED:EMPTY' } if not keys %want_hash;

  ## no critic (ProhibitCommaSeparatedStatements)
  for my $key ( keys %all_keys ) {
    ( push @results, { pass => 0, reason => "KEY:$key:MISSING", } ), next if !exists $got_hash{$key};
    ( push @results, { pass => 0, reason => "KEY:$key:EXCESS", } ),  next if !exists $want_hash{$key};
    ( push @results, { pass => 0, reason => "KEY:$key:NOMATCH", } ), next if !_teq( $got_hash{$key}, $want_hash{$key} );
    push @results, { pass => 1, reason => "KEY:$key:MATCH", };
  }
  return @results;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

DigestX::Multi - Simplify digesting the same content with multiple digesters.

=head1 VERSION

version 0.001000

=head1 SYNOPSIS

  require Digest::SHA3;
  require Digest::SHA;
  require Digest::Whirlpool;

  my $md = DigestX::Multi->new();
  $md->add_digester( 'sha3-512' => Digest::SHA3->new(512) );
  $md->add_digester( 'sha2-512' => Digest::SHA->new(512) );

  # Add byte-strings to all current digesters.
  $md->append_bytes("ByteString", "ByteString", "ByteString");

  # Read IO Handle into all current digesters.
  $md->append_filehandle( $io_handle );

  # Add a digester later that only includes later content
  $md->append_digester( 'whirlpool' => Digest::Whirlpool->new() );

  # Open "path/to/file" and read it into all current digesters.
  $md->append_file( "path/to/file" );

  # Get digest hashes
  # Note: This is intended to be **nondestructive**
  # and internally clones each digester before computing the state.
  my $hashref = $md->digests( format => 'binary' );
   ..         = $md->digests( format => 'hex' );
   ..         = $md->digests( format => 'base64' );


  # true/false
  if ( $md->is_valid( format => 'base64' , digests => $hashref ) ) {
    ...
  }


  # throw exception with details if not valid
  $md->assert_valid( format => 'base64', digests => $hashref );

=head1 DESCRIPTION

C<Digest::Many> aims to simplify common work-flows that are necessary
where-ever somebody might wish to compute multiple digests for the same
content, specifically, with care not to need to open the same files for
multiple passes.

The need to do so arises in especially paranoid situations where the potential
concern involves the potential of any one digest algorithm to be attacked and
have a collision built for, based on the notion that finding one collision for
a given algorithm is hard enough, but it aught to be many orders of magnitude
more difficult to generate a collision that simultaneously collides under two
or more algorithms.

Thus, increasing the number of digest algorithms used exponentially hardens
your attack surface, slightly more so than increasing the digest size of a
single algorithm would achieve.

However, this makes the data less useful for applications where a digest is
used to hide/validate some data using a shared secret, as re-using the same
shared secret with the same content under multiple digests slightly increases
the total number of bits that may relate to the secret, and increases the
chance that the secret may be obtained by breaking a favorable algorithm.

Note that of course, you should not really be using naive secret-prefix or
secret-suffix based message digest authentication, but you should instead use
a L<< proper C<HMAC> algorithm
|https://rdist.root.org/2009/10/29/stop-using-unsafe-keyed-hashes-use-hmac/ >>,
otherwise you're exposing yourself to a L<< length extension attack
|https://www.happybearsoftware.com/you-are-dangerously-bad-at-cryptography.html >>

=head1 METHODS

=head2 C<new>

Create an instance of DigestX::Multi.

  my $instance = DigestX::Multi->new();
  my $instance = DigestX::Multi->new( digesters => { identity => Digest::MD5->new(), } );

This instance starts off with no child digesters, but they can be added via C<add_digester>,
or passed in during construction via C<digesters> as a hash.

=head2 C<add_digester>

Add a new C<Digest> child to this instance.

  $instance->add_digester( my-name => Digest::SHA->new(512) );

C<my-name> must be unique and non-existing.

This method can be called at any time, as long as the user is aware that adding a new C<Digester> late in the cycle
will result in the "Late" digester not including data that was "append"'d previously.

This is deemed a feature.

=head2 C<clone>

Create a new C<Digest::Multi> object where all its child C<digesters> are clones
of this one.

  my $clone = $instance->clone;

=head2 C<reset_digesters>

Call C<reset> on all child digesters.

  $instance->reset_digesters;

=head2 C<append_bytes>

Append all of the strings passed to each child C<digester>.

  $instance->append_bytes("Hello World", "This is also a byte string");

This essentially calls L<< C<< ->add >>|Digest::base/add >> on all child C<digesters> as is.

=head2 C<append_filehandle>

Stream a given C<IO::Handle> until C<EOF> and append its contents efficiently to each current C<digester> child.

  $instance->append_filehandle( path('./foo')->openr_raw );

This will read the given C<filehandle> in C<4KB> chunks and rotate each chunk into each digester via C<append_bytes>

The given C<filehandle> really should be opened in C<raw> mode as everything in C<Digest> really expects C<bytes>.

=head2 C<append_file>

Open the given file in raw mode and iterate its contents through all attached C<digesters>.

  $instance->append_file("./foo");

This will open the given file in C<raw> mode and rotate its contents into the digesters via C<append_filehandle>

=head2 C<digests>

Return a C<HashRef> of C<< identity => digest >>.

  my $hash = $instance->digests(); # format binary assumed

  my $hash = $instance->digests( format => 'binary' );    # use literal binary data
           = $instance->digests( format => 'hex' );       # hex encode digest data
           = $instance->digests( format => 'uuencode' );  # uuencode digest data
           = $instance->digests( format => 'base64' );    # base64(*) encode digest data
           = $instance->digests( encode => sub { } );     # use a custom binary encoding function.

B<*> Base64 data has trailing C<=> padding characters stripped like C<Digest::base>.

B<NOTE:> This method is B<NON DESTRUCTIVE> and can called multiple times to provide the same hash in different encodings.

To facilitate this, C<< ->clone >> is called on each underlying C<digester> before calling their respective C<digest> method.

If a destructive pattern is required, please call C<< ->reset_digesters >> at the appropriate time.

=head2 C<is_valid>

Validate an existing C<HashRef> against accumulated data.

  my $boolean = $instance->is_valid( digests => $hash );  # format binary assumed
  my $boolean = $instance->is_valid( format => binary,  digests => $hash );
              = $instance->is_valid( encode => sub { }, digests => $hash );

See L</VALIDATION>

=head2 C<assert_valid>

Demand accumulated data to be valid using C<HashRef> as a reference, or die explaining all the reasons it isn't.

  $instance->assert_valid( digests => $hash );  # format binary assumed
           ->assert_valid( format => 'binary', digests => $hash );
           ->assert_valid( encode => sub { } , digests => $hash );

See L</VALIDATION>

=head1 VALIDATION

Validation rules are as follows:

=over 4

=item * If the "Digests" hash ( that is, the result of digesting the data )
has no keys, then its a failure, on grounds of the digest configuration being
broken.

=item * If the "Validation" hash ( that is, the hash the digested data is to be
compared against ), then its a failure, on grounds that the validation hash
being empty indicates data corruption.

=item * If a key is present in the "Validation hash" but not in the "Digests"
hash, then its a failure, on grounds that we were unable to locally validate
the data against it.

=item * If a key is present in the "Digests" hash, but not in the "Validation"
hash, then its a failure, on grounds that this may be a hash that is necessary
to confirm no-tampering, and thus, its absence is equivalent to it being
wrong.

=item * If any digest does not match in each hash via corresponding keys, then
its a failure, because any digest failing to match is a sure sign of some kind
of problem occurring.

=item * Finally, if the above conditions are satisfied, assuming the values in
the "Digests" and "Validation" hash are equal for all keys, then its a success.

=back

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( digesters digester encodings ));
for my $wordlist (@stopwords) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
  my $rcfile = path('./perlcritic.rc')->openw_utf8;
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  my $target = path('./misc');
  $target->mkpath if not $target->is_dir;

  my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.047.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Simplify digesting the same content with multiple digesters.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "DigestX-Multi",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "DigestX::Multi",
  "PREREQ_PM" => {
    "Carp" => 0,
    "MIME::Base64" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Builder" => 0,
    "Test::More" => "0.89",
    "Test::Needs" => 0,
    "lib" => 0
  },
  "VERSION" => "0.001000",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/algos/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "MIME::Base64" => 0,
  "Test::Builder" => 0,
  "Test::More" => "0.89",
  "Test::Needs" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
